internal/driver/glfw/window.go

```diff
@@ -404,6 +409,13 @@ func (w *window) Close() {
 	w.closed(w.viewport)
 }
 
+func (w *window) DoClose() {
```

Comment:
Do we really need that to be exported?

Reply:
yes, the idea was that user should explicitly close the window in this new handler

Reply:
IMO a difference between `Close()` and `DoClose()`isn't clear intuitively. I would stick with `TryClose()` and `onCloseTry` naming scheme. @andydotxyz What do you think?

Reply:
DoClose is expected to close window without additional checks so TryClose is exactly opposite name

Reply:
Yep, that what I meant. Don't these names look more intuitive?

Reply:
with the opposite meaning? not sure

Reply:
After some thinking I came up with two alternatives:
1. We have `onClosing func() bool` in window implementation. Now `Close()` closes the window only if `w.onClosing == nil || w.onClosing()`. And no more methods are exported. IMO this way is the most intuitive.
2. We have `onCloseTry func()` in window implementation. It's called when user presses the closing cross or when `w.TryClose()` is called (this is a new exported method). `onCloseTry` decides whether to close the window (via `w.Close()`). IMO this way is more Go-idiomatic, but less intuitive.

Reply:
Regarding the second way. I'm not sure if we need to close window if `TryClose()` is called and no `onCloseTry` handler is defined. But I would choose the first way.

Reply:
I think this got away from me sorry.
I had imagined that "Close()" would ignore the "will" callbacks and just close the window - the developer has told the window to close.
Tapping a close icon or having the app's Quit() called would trigger closing on each window internally, then the developer would be notified, if they requested the hook. In the call-close-to-confirm approach the original Close() method would do just that.

Reply:
The challenge with the `1` option @ababo was that it requires goroutines and messaging - the "willClose" want's to get a boolean return, but the common use-case is to then ask the user... In this case you are blocking a thread whilst waiting for a comfirm callback then passing the value. It made for complicated code.

The alternative, which is basically your `2` above means that the method is a notification and you can confirm with a callback - so threading issues go away.

At least that is what I had thought?...

Reply:
@andydotxyz That makes sense then.

> I'm not sure if we need to close window if `TryClose()` is called and no `onCloseTry` handler is defined.

I would like `Close()` to be called if no `onCloseTry` is defined, otherwise the handler is called. Do you have a preference?

Reply:
@RustamGamidov Are you OK with my naming?

Reply:
not really, let me repeat myself, `Close()` is checking for the additional handler because it's the point where software closing request and window manager close request are coming to. To it looks strange to handle these two cases differently but it's something Fyne team is happy with, i'll step aside

Reply:
@andydotxyz Please make a decision and let's advance with this PR, it's very critical for my project.

Reply:
goroutine is needed anyway, with the current PR this responsibility is moved to the client code.
at least _ShowConfirm_ is deadlocking the app without additional goroutine

Reply:
I think we have crossed lines here, in the original code tapping a window close did not call `Close()` - it calls the internal method `closed()`, which is what the public `Close()` method also does.

If we think that developers calling Close() are doing so to explicitly close the window then `closed()` could have new content, which checks the callback - the main content moved to the `Close()` method and executed if there is no callback. Then the `Close()` does what you'd expect...

Reply:
@andydotxyz can you summarize your vision, including proposed names for the handler and newly exported function?

Reply:
I think that this alternative approach needs only a single new function, named like `SetCloseIntercept(func())` - with the intercept func called only on user action and, if present, will only continue to close if the function calls `Close()`. But this is just me thinking out loud. I have asked if some other contributors have thoughts on this too...

Reply:
@RustamGamidov Can you do as @andydotxyz suggests, i.e. rename `willClose` into `onCloseIntercept`, `SetWillClose()` into `SetCloseIntercept()` and get rid of `DoClose()`? Then the only thing that will remain is an approve from another contributor.

Reply:
yeah... made it if i understood the idea. I'm just confused with how many things are moved to the client code like checking for deadlocks or loops. Especially because there was a comment regarding possible loop

Reply:
It should not be neccessary to do any checking for loops or deadlocks in client code. The simplicity we are looking for is simplicity for the developers using the API.
They should not have to worry about threading at all, that is one of our design principles.

Reply:
then i believe _onCloseIntercepted_ should be called in goroutine but it probably worth to wait for comments from other repo owners because we already used our arguments and most likely we'll stuck arguing

Reply:
I wonder if perhaps the event queue is the right place to put it? Just launching goroutines to work around possible deadlocks seems like it's not fixing the underlying issue? We used to do that with button taps and key presses and it caused a jumble of out-of-order events, so we don't do that any more.